### PR TITLE
Add example demonstrating spike-based Sudoku solver

### PR DESCRIPTION
**Pull request by krachbumm3nte**
A NEST implementation of Sudoku as a constraint satisfaction problem based on (https://github.com/SpiNNakerManchester/IntroLab/tree/master/sudoku)

### CODE REVIEW

**Review by heplesser — CHANGES_REQUESTED** [review]
@krachbumm3nte Thank you for porting this exciting example to NEST. I noticed that some of your scripts reference S Furber and A Rowley as authors. I am therefore wondering to what degree your code is written from scratch and to which degree it is modified from existing code by Furber/Rowley? I believe their original implementation was in PyNN. 

We need to make sure we do the right thing with respect to copyright and licensing. 

I enter this comment as "request changes" to ensure that we won't merge before we have clarified this point.

**Comment by krachbumm3nte** [comment]
@heplesser The actual code shares very little resemblance to the original implementation. The parts I copied are the connection statistics, neuron parameters and the exemplary sudoku puzzles (helpers.py/get_puzzle()). 
I also contacted both of the original authors and got permission to publish my version with their names attached where relevant.

File: pynest/examples/sudoku/sudoku_solver.py [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+
+Credit to the original SpiNNaker implementation of the network used here goes 
+to Steve Furber and Andrew Rowley from the University of Manchester. 
+
+See Also
+---------
+`Original implementation of this Network on SpiNNaker
+<https://github.com/SpiNNakerManchester/IntroLab/tree/master/sudoku>`_
+:doc:`network class <sudoku_net.py>`
+:doc:`script for generating output gifs <plot_progress.py>`
```
Comment (jessica-mitchell) [inline pynest/examples/sudoku/sudoku_solver.py:43]:
```suggestion

:doc:`Script for generating output gifs <plot_progress>`
```

File: pynest/examples/sudoku/sudoku_solver.py [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+:doc:`helper functions <helpers.py>`
```
Comment (jessica-mitchell) [inline pynest/examples/sudoku/sudoku_solver.py:44]:
```suggestion

:doc:`Helper functions <helpers>`
```

File: pynest/examples/sudoku/plot_progress.py [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+Note that the script generates the images individually, storing
+them to disk first, assembling them into a GIF and then,
+by default, deleting the images and folder.
+
+For creating the individual images, PIL is used. Because the GIF creation 
+tool in PIL is faulty, imageio is required for that step.
+
+See Also
+---------
+:doc:`sudoku_solver.py`
```
Comment (jessica-mitchell) [inline pynest/examples/sudoku/plot_progress.py:37]:
```suggestion

:doc:`sudoku_solver`
```

File: pynest/examples/sudoku/plot_progress.py [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+:doc:`helpers.py`
```
Comment (jessica-mitchell) [inline pynest/examples/sudoku/plot_progress.py:38]:
```suggestion
:doc:`helpers`
```

File: pynest/examples/sudoku/helpers.py [hunk 5]
```diff
@@ -0,0 +1,308 @@
...
+font_loc_b = font_manager.findfont(font_prop_b)
+font_bold = ImageFont.truetype(font_loc_b, 18, encoding="unic")
+
+font_prop_it = font_manager.FontProperties(family='sans', style='italic')
+font_loc_it = font_manager.findfont(font_prop_it)
+font_italic = ImageFont.truetype(font_loc_it, 18, encoding="unic")
+
+
+def get_puzzle(puzzle_index):
+    """returns one of 8 sudoku configuration to be solved.
```
Comment (jessica-mitchell) [inline pynest/examples/sudoku/helpers.py:49]:
```suggestion
    """returns one of 8 Sudoku configuration to be solved.
```

File: pynest/examples/sudoku/sudoku_net.py [hunk 6]
```diff
@@ -0,0 +1,247 @@
...
+#
+# NEST is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with NEST.  If not, see <http://www.gnu.org/licenses/>.
+
+r"""Class encapsulating the network for solving sudokus
```
Comment (jessica-mitchell) [inline pynest/examples/sudoku/sudoku_net.py:22]:
```suggestion
r"""Class encapsulating the network for solving Sudoku
```

File: pynest/examples/sudoku/README.rst [hunk 7]
```diff
@@ -0,0 +1,3 @@
+nest Sudoku
+========
+A PyNest implementation of Sudoku as a constraint satisfaction problem
```
Comment (jessica-mitchell) [inline pynest/examples/sudoku/README.rst:3]:
```suggestion
A PyNEST implementation of Sudoku as a constraint satisfaction problem
```

**Review by jessica-mitchell — CHANGES_REQUESTED** [review]
Thanks for this contribution @krachbumm3nte, really cool example! I noted a few things to fix up with the formatting and such. Please make sure to fix conflicts first before applying reviewer changes.

File: pynest/examples/sudoku/sudoku_net.py [hunk 8]
```diff
@@ -0,0 +1,247 @@
...
+
+Notes
+------
+The functionality of the network lies entirely within the inhibitory
+connections between neuron populations. each population of size ``n_digit``
+encodes a digit between 1 and 9 in one of the 81 cells and has outgoing 
+inhibitory connections to several other populations. Namely all populations
+coding for the same digit in the same row, column and 3x3 box of the sudoku 
+field, since a given digit can only ever occur once in any of those three. It 
+also inhibits all populations in the same cell which code for different digits
```
[the remark this answers is not in the document]
Comment (krachbumm3nte) [reply pynest/examples/sudoku/sudoku_net.py]: This one does not seem appropriate but the original phrasing is unfortunate. I changed this to "[...] which encode different digits" in [my previous commit](https://github.com/krachbumm3nte/nest-simulator/commit/567549446560fef5f9b0ee5d234eeb521dc39737) (I spotted the "add a suggestion" button too late).

File: doc/userdoc/examples/index.rst [hunk 9]
```diff
@@ -128,4 +128,9 @@ PyNEST examples
 
    ../auto_examples/hpc_benchmark
 
+.. toctree::
+   :maxdepth: 1
+   :caption: Sudoku
 
+   ../auto_examples/sudoku/sudoku_net
+   ../auto_examples/sudoku/sudoku_solver
```
[the remark this answers is not in the document]
Comment (krachbumm3nte) [reply doc/userdoc/examples/index.rst]: I'm not entirely sure what to make of the changes shown here - the first two lines appear to be identical. But the new file in doc/htmldoc/examples/index.rst is up to date with this, so I am ready to mark this as resolved.

File: pynest/examples/sudoku/sudoku_net.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (jessica-mitchell) [reply pynest/examples/sudoku/sudoku_net.py]: Ok, that sounds better thanks

File: doc/userdoc/examples/index.rst (hunk 9 above)
[the remark this answers is not in the document]
Comment (jessica-mitchell) [reply doc/userdoc/examples/index.rst]: The plot_progress python script is missing from the list -and in my view I don't see that it's added.

[the remark this answers is not in the document]
Comment (krachbumm3nte) [reply doc/userdoc/examples/index.rst]: Apologies for that. [I have added it.](https://github.com/krachbumm3nte/nest-simulator/commit/a18676630c470306e6f38b2bcf0612b4cb76e825)

File: pynest/examples/sudoku/plot_progress.py [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+by default, deleting the images and folder.
+
+For creating the individual images, PIL is used. Because the GIF creation 
+tool in PIL is faulty, imageio is required for that step.
+
+See Also
+~~~~~~~~
+
+:doc:`sudoku_solver`
+:doc:`helpers`
```
Comment (jessica-mitchell) [inline pynest/examples/sudoku/plot_progress.py:39]:
```suggestion

:doc:`helpers`
```

Comment (krachbumm3nte) [reply pynest/examples/sudoku/plot_progress.py:40]: Thank you, I am very much open to that. Let me know if you need my help.

**Review by gtrensch — APPROVED** [review]
Thank you for porting this example to NEST. :+1: 
I approve the PR, but also would have a minor comment: While testing the example, I went into a dependency issue related to the Python imaging library `PLI`. Discussed this with @terhorstd, and we would like to replace `PLI` with a `matplotlib` solution at a later time. We opened #2441 to follow up on this.

**Pull request merged**